Micro-optimize Active Record connection checkin

byroot opened the pull request:
Ref: https://github.com/rails/rails/issues/55728

Starting in 7.2, since we no longer pin Active Record connection for the entirely of the request or job cycle, `Pool#checkout` and `Pool#checkin` overhead has to be paid for every query. It's not a whole lot once you consider a query will need to do a network hop, but we should still try to keep that overhead low.

This PR contains many different small optimizations, many of them too small to show a clear gain on a benchmark, but together they add up, see individual commits. (FYI: @cmaion).


Which brings up back to about half-way between 7.1 and `main`:

```
ruby 3.4.6 (2025-09-16 revision dbd83256b1) +YJIT +PRISM [arm64-darwin24]
Warming up --------------------------------------
          7-1-stable     3.219k i/100ms
Calculating -------------------------------------
          7-1-stable     88.512k (± 4.1%) i/s   (11.30 μs/i) -    444.222k in   5.027300s

Comparison:
     ar-fast-checkin:    79704.1 i/s
          7-1-stable:    88512.0 i/s - 1.11x  faster
                main:    70279.4 i/s - 1.13x  slower
```

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", path: "."
  gem "sqlite3"
  gem "benchmark-ips"
end

require "active_record/railtie"
require "benchmark/ips"

# This connection will do for database-independent bug reports.
ENV["DATABASE_URL"] = "sqlite3::memory:"

class TestApp < Rails::Application
  config.load_defaults Rails::VERSION::STRING.to_f
  config.eager_load = false
  config.logger = Logger.new(nil)
  config.log_level = :info
  config.secret_key_base = "secret_key_base"

  config.active_record.encryption.primary_key = "primary_key"
  config.active_record.encryption.deterministic_key = "deterministic_key"
  config.active_record.encryption.key_derivation_salt = "key_derivation_salt"
end
Rails.application.initialize!


ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
  end
end

ActiveRecord::Base.release_connection unless Rails::VERSION::STRING < "7.2"

class Post < ActiveRecord::Base
end

BRANCH = `git rev-parse --abbrev-ref HEAD`.strip

Benchmark.ips do |x|
  x.report(BRANCH) { Post.count }
  x.save!("/tmp/bench-ar-data")
  x.compare!(order: :baseline)
end
```

There might be some more optimizations to perform, but I think this is now large enough to be worth merging.

byroot committed:
ActiveSupport::Notifications leverage `...` delegation

Recent Rubies are able to perform `...` delegation with 0 allocations
we should use it when possible.

byroot commented on activerecord/lib/active_record/runtime_registry.rb:
```diff
@@ -7,76 +7,80 @@ module ActiveRecord
-  end
-  ActiveRecord::RuntimeRegistry.sql_runtime += runtime
-end
+# ActiveSupport::Notifications.monotonic_subscribe("sql.active_record") do |name, start, finish, id, payload|
```
No, it's just temporary to measure how much we can save in a case where this is no active `sql.active_record` subscribers. The answer is quite a lot, but then how realistic such a scenario is, I'm not sure.

I suspect it's very common for apps to have some APM or similar so probably at least one subscriber here.

The root of the problem is that `ActiveSupport::Notifications` really isn't as cheap as I'd like, but it already has been optimized quite far, I doubt we can optimize it much more while keep perfect compatibility.

byroot commented on activerecord/lib/active_record/explain_subscriber.rb:
```diff
@@ -29,6 +29,10 @@ def ignore_payload?(payload)
         !payload[:sql].match?(EXPLAINED_SQLS)
     end
 
+    def silenced?(event)
+      !ExplainRegistry.collect?
+    end
```
Better than silencing it, we should just only subscribe on the event when we're collecting.

This means we'd flush the AS::N cache, but I think it's OK as you'd only realistically do this in development.

byroot committed:
Refactor preprocess_query to call `write_query?` only once

This saves having to match the same regexp twice in a row.

Ideally we wouldn't even use a regexp in most cases, and instead
rely on the higher level hints, e.g. if `select_all` was used
we should assume it's a read.

rafaelfranca commented:
> My problem is that this is a lot of small optimizations that are within the error margin when benchmarked individually, so hard to justify them... Hence not to sure how to ship that.

I'd not worry with that. If the aggregation of patches make the code faster, I'd just apply them all together (in individual commits)

byroot committed:
ActiveSupport::Callbacks add a fast path when there is no callbacks

When initially defined, the `_run_<name>_callbacks` method is a noop.

Only once a callback has been registers, we swap the implementation
for the real one.

This saves a little bit of performance for hooks points that are
rarely used or only used in development.

byroot committed:
Refactor ExplainRegistry to only be subscribed once used

Realistically, it's only ever going to be invoked in development,
hence it is wasteful to have a notification subscriber that is
a noop.

Instead we can delay the subscription to when the explain collection
is enabled for the first time, meaning except for a few rare cases
this code won't even be loaded in production environments.

byroot committed:
ActiveRecord::AbstractAdapter: don't use callbacks for normal operations

Callbacks have a fairly significant overhead, and don't really make the
code easier to follow. They are useful to allow third party code
to hook into the connections lifecycle, but there isn't a big
reason for internal code to use them rather than a regular method call.

byroot committed:
ActiveRecord LeaseRegistry: skip locking on MRI

Thanks to the GVL, we don't need that mutex when running on MRI.
This remove a little bit of overhead on every connection checkout and checkin.

byroot committed:
Micro optimize ActiveRecord::RuntimeRegistry

Accessing `IsolatedExecutionState` is a bit costly,
hence it's prefereable to minimize accesses.

By moving all 4 RuntimeRegistry metrics in a consolidated object,
we now only access the `IsolatedExecutionState` once per query
while before it was up to 8 times.

byroot committed:
Micro-optimize ActiveSupport::Notifications

- Adds a fastpath in `iterate_guarding_exceptions` for when there's
only a single subscriber. In such case we don't need any fancy exception
handling.

- Merge the `groups_for` and `silenceable_group_for` caches, as to
fetch both records with a single lookup.

- Return a null object in `build_handle` if there are no subscribers

byroot commented:
Alright, I've cleaned up the git history, the benchmark is now:

```
ruby 3.4.6 (2025-09-16 revision dbd83256b1) +YJIT +PRISM [arm64-darwin24]
Warming up --------------------------------------
          7-1-stable     3.219k i/100ms
Calculating -------------------------------------
          7-1-stable     88.512k (± 4.1%) i/s   (11.30 μs/i) -    444.222k in   5.027300s

Comparison:
     ar-fast-checkin:    79704.1 i/s
          7-1-stable:    88512.0 i/s - 1.11x  faster
                main:    70279.4 i/s - 1.13x  slower
```

I think there is a bit more perf to squeeze in AS:N, but I'd rather not slip on these gains and merge soon.

I also removed the interlock patch given Matthew merged https://github.com/rails/rails/pull/55753

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:
```diff
@@ -179,21 +179,30 @@ def clear(connection)
+        # Thanks to the GVL, the LeaseRegistry doesn't need to be synchronized on MRI
+        class LeaseRegistry < WeakThreadKeyMap # :nodoc:
+          def [](context)
+            super || (self[context] = Lease.new)
```
Is this safe just because we know that we _are_ `context`, and therefore even if we stall during `Lease.new`, we can't possibly be racing someone else?

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -592,8 +592,10 @@ def affected_rows(raw_result)
-          mark_transaction_written_if_write(sql)
+          if write_query?(sql)
+            ensure_writes_are_allowed(sql)
+            mark_transaction_written
```
Should we just inline both of these? They're both basically `x if y`, which was probably more method-worthy when they were called separately from each adapter etc? 🤷🏻‍♂️

byroot replied:
I suppose we could yes. More importantly we should pass down whether we expect a read or write query, e.g. if we're called from `select_all`, we should assume it's a read query.

That regexp is about 1.5% on the profile AFAICT.

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:
```diff
@@ -330,8 +328,12 @@ def expire(update_idle = true) # :nodoc:
+          _run_checkin_callbacks do
+            @idle_since = Process.clock_gettime(Process::CLOCK_MONOTONIC) if update_idle
+            @owner = nil
+            enable_lazy_transactions!
```
Not something to change here, but consolidating like this draws my eye to the fairly arbitrary distribution of which "return to neutral state" things happen during checkin, and which during checkout.

matthewd commented on activerecord/lib/active_record/explain_registry.rb:
```diff
@@ -24,6 +69,11 @@ def initialize
     end
 
+    def start
+      Subscriber.ensure_subscribed
```
Nice!

matthewd commented on activesupport/lib/active_support/notifications/fanout.rb:
```diff
@@ -184,25 +187,25 @@ def build_event(name, id, payload)
-            s.map(&:delegate)
-          end
-        end
+      def group_listeners(listeners)
```
```suggestion
      def group_listeners(listeners) # :nodoc:
```
